Vizzuality/widget-editor
# Support rendering and creating custom colour schemes

This PR adds support for the custom schemes embedded in the widgets, the schemes embedded in the widgets that are not officially supported by the host app and gives the possibility to the user to create their own custom colour scheme.

<p align="center">
<img width="1424" alt="The user has created a rainbow custom scheme" src="https://user-images.githubusercontent.com/6073968/89548866-7eab2880-d7ff-11ea-91d6-5c6242015fc7.png">
</p>

To fully understand the extent of the changes, here is the list of all the cases supported:
1. The widget-editor is instantiated without any widget
2. The widget-editor is instantiated with a widget that doesn't have an embedded scheme
3. The widget-editor is instantiated with a widget that has an embedded scheme and that scheme is supported by the host app (i.e. its name can be found in the list passed to the `schemes` prop)
4. The widget-editor is instantiated with a widget that has an embedded scheme and that scheme is _not_ supported by the host app
5. The widget-editor is instantiated with a widget that has an embedded scheme, tagged as custom (i.e. its name is “user-custom”)

In cases **1.** and **2.**, the default scheme applied to the widget is the first scheme of the list passed to the `schemes` prop, or the default scheme of the editor (called “Resource Watch”) if the host app doesn't pass schemes to the editor.

In case **3.**, the default scheme applied to the widget is the embedded scheme.

In cases **4.** and **5.**, the default scheme applied to the widget is also the embedded scheme, but no matter the embedded scheme name, it will be displayed as “Custom” in the UI.

In addition, the user will have the possibility to create its own custom scheme.

In cases **1.**, **2.** and **3.**, the user will have to click the button “Create custom scheme” to create their own scheme. The custom scheme will inherit the previously selected scheme's colours.

In cases **4.** and **5.**, a “Custom” option will already be present in the UI, as it represents the embedded scheme. The user will be able to directly edit this scheme.

It's important to note that in case **4.**, the scheme will have a name different from “user-custom”. As soon as the user starts to customise the scheme, its name will be switched to “user-custom”.

## Testing instructions

I would suggest testing the different cases laid out above:
1. Instantiate the editor with this dataset: `852f2275-91a8-4500-9f10-89880dc53f22`
2. Instantiate with this widget: `b867b473-d4f9-4105-b87c-df76caaaf358` (same dataset)
3. Instantiate with this widget: `c1d7cb83-5f50-43eb-8ea0-6f7ecd4e1b69` (same dataset)**(*)**
4. Instantiate with this widget: `c1d7cb83-5f50-43eb-8ea0-6f7ecd4e1b69` (same dataset) and remove the “default” scheme from the playground's `SCHEMES` variable (found here: `src/playground/widget-editor/src/components/editor/component.js`)
5. Instantiate with this widget: `5a929bd2-8aeb-491d-8814-05d6f9519534` (same dataset)

In addition, I would suggest customising a scheme and checking the editor's payload.

**(*)** The widget's embedded scheme is called “default”. As such, the “default” scheme will be applied by default. Nevertheless, “default” is already the first scheme in the playground's `SCHEMES` variable (found here: `src/playground/widget-editor/src/components/editor/component.js`). I would suggest moving the playground's “default” scheme to another position to correctly check if the embedded scheme is restored.

There's also a catch here, that shows the limitations of using the scheme's name as a unique identifier: the widget's embedded scheme, while called “default” and understood as the same “default” scheme of the `SCHEMES` list, actually defines a different colour scheme than the one in `SCHEMES`. You can see that the main colour of the embedded scheme is green, while the one of the `SCHEMES` list is blue.

## Pivotal Tracker

[Task](https://www.pivotaltracker.com/story/show/172822853).
